packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -1,242 +1,273 @@
+          this.#isDetectionEnabledFromPreferences !== useTokenDetection;
+
+        this.#selectedAddress = newSelectedAddress;
+        this.#isDetectionEnabledFromPreferences = useTokenDetection;
+
+        if (
+          useTokenDetection &&
+          (isSelectedAddressChanged || isDetectionChangedFromPreferences)
+        ) {
+          await this.detectTokens();
+        }
+      },
+    );
+
+    this.messagingSystem.subscribe(
+      'NetworkController:networkDidChange',
+      async ({ selectedNetworkClientId }) => {
+        this.#networkClientId = selectedNetworkClientId;
+        const newChainId = this.#getCorrectChainId(selectedNetworkClientId);
+        const isChainIdChanged = this.#chainId !== newChainId;
+        this.#chainId = newChainId;
+
+        this.#isDetectionEnabledForNetwork =
+          isTokenDetectionSupportedForNetwork(newChainId);
+
+        if (this.#isDetectionEnabledForNetwork && isChainIdChanged) {
+          await this.detectTokens();
+        }
+      },
+    );
   }
 
   /**
    * Start polling for detected tokens.
    */
   async start() {
-    this.configure({ disabled: false });
-    await this.startPolling();
+    this.#disabled = false;
+    await this.#startPolling();
   }
 
   /**
    * Stop polling for detected tokens.
    */
   stop() {
-    this.configure({ disabled: true });
-    this.stopPolling();
+    this.#disabled = true;
+    this.#stopPolling();
   }
 
-  private stopPolling() {
-    if (this.intervalId) {
-      clearInterval(this.intervalId);
+  #stopPolling() {
+    if (this.#intervalId) {
+      clearInterval(this.#intervalId);
     }
   }
 
   /**
    * Starts a new polling interval.
-   *
-   * @param interval - An interval on which to poll.
    */
-  private async startPolling(interval?: number): Promise<void> {
-    interval && this.configure({ interval }, false, false);
-    this.stopPolling();
+  async #startPolling(): Promise<void> {
+    if (this.#disabled) {
+      return;
+    }
+    this.#stopPolling();
     await this.detectTokens();
-    this.intervalId = setInterval(async () => {
+    this.#intervalId = setInterval(async () => {
       await this.detectTokens();
-    }, this.config.interval);
+    }, this.getIntervalLength());
   }
 
-  private getCorrectChainId(networkClientId?: NetworkClientId) {
-    if (networkClientId) {
-      return this.getNetworkClientById(networkClientId).configuration.chainId;
-    }
-    return this.config.chainId;
+  #getCorrectChainId(networkClientId?: NetworkClientId) {
+    const {
+      configuration: { chainId },
+    } = this.messagingSystem.call(
+      'NetworkController:getNetworkClientById',
+      networkClientId ?? this.#networkClientId,
+    );
+    return chainId;
   }
 
-  _executePoll(
+  async _executePoll(
     networkClientId: string,
     options: { address: string },
   ): Promise<void> {
-    return this.detectTokens({
+    if (this.#disabled) {
+      throw new Error(
```

Comment:
It looks like `detectTokens` already doesn't do anything if `disabled` is true, so do we need to throw an error? We aren't doing this in any other controllers' `_executePoll` methods, so this would be a first.

Reply:
I added this error because `_executePoll` is the only non-private method aside from `detectTokens` itself where the `disabled` flag blocks a request. But a runtime error is unexpected and disruptive, I can remove it.

I would like to keep the if-block early exit, though, to make the role of `disabled` explicit for anyone adding code to this method in the future.

Reply:
This method isn't really meant to be public, hence the leading underscore. It's only public because that's how we were able to get mixins to work. I wouldn't worry too much about it being called directly from anywhere other than the mixin.

Getting rid of the error seems like a good idea to me too; it would clutter up the dev console if we left it in.

Reply:
Thanks for explaining this. Error removed here: 1552041bfd080a83b7cd95f86d38be074c9b6235